### DISCUSSION TITLE
What makes it different?

### DISCUSSION THREAD

**Discussion by petershaw** [hummingbird-project/hummingbird#150] [Q&A] [answered]
Comparing to Vapor, what makes Hummingbird different? When to use it, and when to avoid? What are the strength of it and what problem target does it solve. 

I am very interested in new swift frameworks for the server land. Dokumentation needs a "motivation" section and a guide how to contribute. 

Thanks,
ps.

**Comment by adam-fowler** [comment, marked as answer]
Vapor provides almost everything you need in one package (HTTP2, TLS, Authentication, OTP, caching, compression, validation, web sockets etc). It's a great one stop shop. Though this makes it a very heavy package which bulks out your executable, and increases compile time. It also provides wrappers for many other packages so they can be used in the expected manner. It is quite prescriptive (admittedly there are hummingbird redis and fluent wrappers but I'm considering slimming them down).

Hummingbird is designed to be more slimline. A simple hummingbird app, contains pretty much only the server and router. I've used it to set up a web server, serving local files for testing static web sites (https://github.com/adam-fowler/swift-web), you could use it to setup an http proxy server (see example https://github.com/hummingbird-project/hummingbird-examples/tree/main/proxy-server). Given the difference in size, hummingbird is probably more suited to building lambdas than vapor (using https://github.com/hummingbird-project/hummingbird-lambda) where package size matters. 

While it is more suited to smaller projects it is still capable of supporting larger projects. If you do need additional features there are packages that cover most features (authentication, tls, http2, web sockets, offloading tasks etc).

Hummingbird is faster than Vapor. The core HTTP server can serve almost double the number of requests as Vapor (see the latest TechEmpower [results](https://www.techempower.com/benchmarks/#section=data-r21&test=plaintext&l=v2qiv3-6bj)). This is most likely due to Vapor including more features by default but I have spent a considerable amount of time trying to optimise the framework for speed.

There is a [`CONTRIBUTING.md`](https://github.com/hummingbird-project/hummingbird/blob/main/CONTRIBUTING.md) guide here. Although I think I need to expand that a bit.

**Comment by cemery50** [comment]
It would be interesting for an ai to spin up functions as needed with Hummingbird as an atomic start point...maybe it could allocate responsibilities to a hybrid cloud enviornment based upon needs
Does hummingbird work with the atomic version of Fedora ?

**Reply by adam-fowler** [reply]
You could probably use the static linux SDK to build a version that would run on Fedora. See https://www.swift.org/documentation/articles/static-linux-getting-started.html
